passes/opt/muxpack.cc

```diff
@@ -0,0 +1,270 @@
+ *  WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ *  MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ *  ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ *  WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ *  ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ *  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *
+ */
+
+#include "kernel/yosys.h"
+#include "kernel/sigtools.h"
+
+USING_YOSYS_NAMESPACE
+PRIVATE_NAMESPACE_BEGIN
+
+struct MuxpackWorker
+{
+	Module *module;
+	SigMap sigmap;
+
+	int mux_count, pmux_count;
+
+	pool<Cell*> remove_cells;
+
+	dict<SigSpec, Cell*> sig_chain_next;
+	dict<SigSpec, Cell*> sig_chain_prev;
+	pool<SigBit> sigbit_with_non_chain_users;
+	pool<Cell*> chain_start_cells;
+	pool<Cell*> candidate_cells;
+
+	void make_sig_chain_next_prev()
+	{
+		for (auto wire : module->wires())
+		{
+			if (wire->port_output || wire->get_bool_attribute("\\keep")) {
+				for (auto bit : sigmap(wire))
+					sigbit_with_non_chain_users.insert(bit);
+			}
+		}
+
+		for (auto cell : module->cells())
+		{
+			if (cell->type.in("$mux", "$pmux") && !cell->get_bool_attribute("\\keep"))
+			{
+				SigSpec a_sig = sigmap(cell->getPort("\\A"));
+				SigSpec b_sig;
+				if (cell->type == "$mux")
+					b_sig = sigmap(cell->getPort("\\B"));
+				SigSpec y_sig = sigmap(cell->getPort("\\Y"));
+   
+				if (sig_chain_next.count(a_sig))
+					for (auto a_bit : a_sig.bits())
+						sigbit_with_non_chain_users.insert(a_bit);
+				else {
+					sig_chain_next[a_sig] = cell;
+					candidate_cells.insert(cell);
+				}
+
+				if (!b_sig.empty()) {
+					if (sig_chain_next.count(b_sig))
+						for (auto b_bit : b_sig.bits())
+							sigbit_with_non_chain_users.insert(b_bit);
+					else {
+						sig_chain_next[b_sig] = cell;
+						candidate_cells.insert(cell);
+					}
+				}
+
+				sig_chain_prev[y_sig] = cell;
+				continue;
+			}
+
+			for (auto conn : cell->connections())
+				if (cell->input(conn.first))
+					for (auto bit : sigmap(conn.second))
+						sigbit_with_non_chain_users.insert(bit);
+		}
+	}
+
+	void find_chain_start_cells()
+	{
+		for (auto cell : candidate_cells)
+		{
+			log_debug("Considering %s (%s)\n", log_id(cell), log_id(cell->type));
+
+			SigSpec a_sig = cell->getPort("\\A");
+			if (cell->type == "$mux") {
+				SigSpec b_sig = cell->getPort("\\B");
+				if (sig_chain_prev.count(a_sig) + sig_chain_prev.count(b_sig) != 1)
+					goto start_cell;
+
+				if (!sig_chain_prev.count(a_sig))
+					a_sig = b_sig;
+			}
+			else if (cell->type == "$pmux") {
+				if (!sig_chain_prev.count(a_sig))
+					goto start_cell;
+			}
+			else log_abort();
+
+			{
+				for (auto bit : a_sig.bits())
+					if (sigbit_with_non_chain_users.count(bit))
+						goto start_cell;
+
+				Cell *c1 = sig_chain_prev.at(a_sig);
+				Cell *c2 = cell;
+
+				if (c1->getParam("\\WIDTH") != c2->getParam("\\WIDTH"))
+					goto start_cell;
```

Comment:
Shouldn't this already be guaranteed by the fact that c1.Y == c2.A? However, if this is run after freduce then you want to strip zero extension from A/B/Y before creating the database entries and before the lookups. Here's a test case (use with `proc; opt; freduce; opt`):

```
module test ( 
        input wire [1:0] s,
        input wire a, b, c, d,
        output reg [3:0] o
);
        always @* begin
                o = {4{a}};
                if (s == 0) o = {3{b}};
                if (s == 1) o = {2{c}};
                if (s == 2) o = d;
        end
endmodule
```

Reply:
Ooh `freduce` is a funky one. Basically, it replaces redundant wires on the driving cell.`Y` with dummy wires, and on the sink cell's input, it un-redundant-ises those previous bits to get the original signal back.

For example, driving.Y might have been `{a,a,a,a}` which `freduces` transforms into `{dummy1,dummy2,dummy3,a}`. sink.A would then, instead of just receiving `Y`, would recreate it by doing `{Y[0],Y[0],Y[0],Y[0]}`.

Given that we rely on driving.Y == sink.A in order to follow the chain of muxes, wouldn't using `freduce` break this flow?

Reply:
Do you have a MCVE that I can use to reproduce the case where freduce prevents muxpack from working as desired?

Reply:
I used your exact testcase here: https://github.com/YosysHQ/yosys/blob/4ddc0354c1cc61f2e2b3f15cc341fd277c710e89/tests/various/muxpack.v#L170-L181 on script (currently commented out): https://github.com/YosysHQ/yosys/blob/4ddc0354c1cc61f2e2b3f15cc341fd277c710e89/tests/various/muxpack.ys#L184-L199
which gives:
![before](https://user-images.githubusercontent.com/6740044/60047879-ab45c480-967f-11e9-8446-30a7c635cc7a.png)
which seems to be quite messy to analyse, no?

Reply:
Thanks. I'll have a closer look.